derive/src/trait_bounds.rs

```diff
@@ -51,10 +51,21 @@ pub fn make_where_clause<'a>(
 
     let types = collect_types_to_bind(input_ident, data, &ty_params_ids)?;
 
-    types.into_iter().for_each(|ty| {
-        where_clause
-            .predicates
-            .push(parse_quote!(#ty : ::scale_info::TypeInfo + 'static))
+    types.into_iter().for_each(|(ty, is_compact)| {
+        // Compact types need extra bounds, T: HasCompact and <T as
+        // HasCompact>::Type: TypeInfo + 'static
+        if is_compact {
+            where_clause
+                .predicates
+                .push(parse_quote!(#ty : ::scale::HasCompact));
+            where_clause
+                .predicates
+                .push(parse_quote!(<#ty as ::scale::HasCompact>::Type : ::scale_info::TypeInfo + 'static));
```

Comment:
I think our use of `::scale` is becoming a problem. ink! imports `parity-scale-codec` crate as `scale` but from what I remember Substrate imports it as `codec`. We need a solution that works for both. Probably Basti's https://crates.io/crates/proc-macro-crate could help here. For now I am fine and we can fix this later.

Reply:
this also seem to break primitive-types: https://github.com/paritytech/parity-common/pull/519